distogram/__init__.py

```diff
@@ -1,240 +1,203 @@
+            if h.diffs[i - 1] < h.min_diff:
+                h.min_diff = h.diffs[i - 1]
+
         if i < len(h.bins) - 1:
-            diff = _weighted_diff(h, h.bins[i+1], h.bins[i])
             if h.diffs[i] == h.min_diff:
-                h.diffs[i] = diff
                 update_min = True
-            else:
-                h.diffs[i] = diff
-                if h.diffs[i] < h.min_diff:
-                    h.min_diff = h.diffs[i]
+
+            h.diffs[i] = _weighted_diff(h, h.bins[i + 1], h.bins[i])
+            if h.diffs[i] < h.min_diff:
+                h.min_diff = h.diffs[i]
 
         if update_min is True:
             h.min_diff = min(h.diffs)
 
+    return
+
 
-def _trim(h, index):
-    bins = h.bins
-    while len(bins) > h.bin_count:
+def _trim(h: Distogram) -> Distogram:
+    while len(h.bins) > h.bin_count:
         if h.diffs is not None:
-            min_diff = h.min_diff
-            i = h.diffs.index(min_diff)
+            i = h.diffs.index(h.min_diff)
         else:
-            min_diff = None
-            i = None
-            for index, value in enumerate(h.bins):
-                if index > 0:
-                    diff = _weighted_diff(h, h.bins[index], h.bins[index-1])
-                    if min_diff is None or diff < min_diff:
-                        min_diff = diff
-                        i = index - 1
-
-        bins[i] = (
-            (bins[i][0]*bins[i][1] + bins[i+1][0]*bins[i+1][1])
-            / (bins[i][1] + bins[i+1][1]),
-            bins[i][1] + bins[i+1][1]
-        )
-        del bins[i+1]
+            diffs = [
+                (i - 1, _weighted_diff(h, b, h.bins[i - 1]))
+                for i, b in enumerate(h.bins[1:], start=1)
+            ]
+            i, _ = min(diffs, key=itemgetter(1))
+
+        v1, f1 = h.bins[i]
+        v2, f2 = h.bins[i + 1]
+        h.bins[i] = (v1 * f1 + v2 * f2) / (f1 + f2), f1 + f2
+        del h.bins[i + 1]
+
         if h.diffs is not None:
             del h.diffs[i]
             _update_diffs(h, i)
             h.min_diff = min(h.diffs)
 
-    h.bins = bins
     return h
 
 
-def _trim_in_place(h, value, count, i):
-    bins = h.bins
-
-    bins[i] = (
-        (bins[i][0]*bins[i][1] + value*count)
-        / (bins[i][1] + count),
-        bins[i][1] + count
-    )
-
-    h.bins = bins
+def _trim_in_place(h: Distogram, value: float, c: int, i: int):
+    v, f = h.bins[i]
+    h.bins[i] = (v * f + value * c) / (f + c), f + c
     _update_diffs(h, i)
     return h
 
 
-def _compute_diffs(h):
-    diffs = []
-    bins = h.bins
-    h.min_diff = None
-    for index in range(1, len(bins)):
-        diff = _weighted_diff(h, h.bins[index], h.bins[index-1])
-        diffs.append(diff)
-        if h.min_diff is None or diff < h.min_diff:
-            h.min_diff = diff
+def _compute_diffs(h: Distogram) -> List[float]:
+    diffs = [
+        _weighted_diff(h, b2, b1)
+        for b1, b2 in zip(h.bins[:-1], h.bins[1:])
+    ]
+    h.min_diff = min(diffs)
```

Comment:
We compute the diffs in a single list-comprehension by offsetting the bins. Then at the end, we set the min-diff only once.